server/internal/server/vector_stores.go

```diff
@@ -0,0 +1,321 @@
+	var protos []*v1.VectorStore
+	for _, c := range cs {
+		cm, err := s.store.ListCollectionMetadataByCollectionID(c.CollectionID)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "list collection metadata: %s", err)
+		}
+		protos = append(protos, toVectorStoreProto(c, cm))
+	}
+	first := ""
+	last := ""
+	if len(protos) > 0 {
+		first = protos[0].Id
+		last = protos[len(protos)-1].Id
+	}
+	return &v1.ListVectorStoresResponse{
+		Object:  vectorStoreObject,
+		Data:    protos,
+		FirstId: first,
+		LastId:  last,
+		HasMore: hasMore,
+	}, nil
+}
+
+// UpdateVectorStore updates a vector store.
+func (s *S) UpdateVectorStore(
+	ctx context.Context,
+	req *v1.UpdateVectorStoreRequest,
+) (*v1.VectorStore, error) {
+	userInfo, err := s.extractUserInfoFromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	id, err := getCollectionID(req.VectorStoreId, true)
+	if err != nil {
+		return nil, err
+	}
+
+	c, err := s.store.GetCollectionByCollectionID(userInfo.ProjectID, id)
+	if err != nil {
+		if errors.Is(err, gorm.ErrRecordNotFound) {
+			return nil, status.Errorf(codes.NotFound, "collection %q not found", req.VectorStoreId)
+		}
+		return nil, status.Errorf(codes.Internal, "get collection: %s", err)
+	}
+	if c.ProjectID != userInfo.ProjectID {
+		return nil, status.Errorf(codes.PermissionDenied, "collection %q not found", req.VectorStoreId)
+	}
+
+	// update collection metadata
+	cms, err := s.store.ListCollectionMetadataByCollectionID(c.CollectionID)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "list collection metadata: %s", err)
+	}
+	cur := make(map[string]*store.CollectionMetadata)
+	for _, cm := range cms {
+		cur[cm.Key] = cm
+	}
+	for k, v := range req.Metadata {
+		found, ok := cur[k]
+		if !ok {
+			cm := &store.CollectionMetadata{
+				CollectionID: c.CollectionID,
+				Key:          k,
+				Value:        v,
+			}
+			if err := s.store.CreateCollectionMetadata(cm); err != nil {
+				return nil, status.Errorf(codes.Internal, "create collection metadata: %s", err)
+			}
+		} else if found.Value != v {
+			found.Value = v
+			if err := s.store.UpdateCollectionMetadata(found); err != nil {
+				return nil, status.Errorf(codes.Internal, "update collection metadata: %s", err)
+			}
+		}
+	}
+	for _, cm := range cms {
+		if _, ok := req.Metadata[cm.Key]; !ok {
+			if err := s.store.DeleteCollectionMetadata(cm.CollectionID, cm.Key); err != nil {
+				return nil, status.Errorf(codes.Internal, "delete collection metadata: %s", err)
+			}
+		}
+	}
+	cms, err = s.store.ListCollectionMetadataByCollectionID(c.CollectionID)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "list collection metadata: %s", err)
+	}
+
+	// update collection
+	if req.Name != "" && req.Name != c.Name {
+		if err := s.vstoreClient.UpdateVectorStoreName(ctx, c.Name, req.Name); err != nil {
+			return nil, status.Errorf(codes.Internal, "update vector store: %s", err)
+		}
+		c.Name = req.Name
+	}
+	if req.ExpiresAfter != nil {
```

Comment:
There is no way to clear `ExpiresAfter` once it is set.

It is not clear from the API reference page, but it might make sense to unset `ExpiresAfter` it is not set in the request. We could follow the same principle for metadata (and return an error if name is not specified).

Reply:
Setting `days` to 0 is equivalent to unset, I think?

Reply:
I'm not sure. Maybe we should test with OpenAI API, but I expect that setting `days` to 0 will return an invalid argument error.

Reply:
Tested.

<img width="1824" alt="image" src="https://github.com/llm-operator/vector-store-manager/assets/8353550/93155088-a3bc-4c74-b3b7-4df2bb92c581">


Response:

```
{
  "id": "vs_v9WodrpusSGehNzFAT1S1Gcp",
  "object": "vector_store",
  "name": null,
  "status": "completed",
  "usage_bytes": 0,
  "created_at": 1718321233,
  "file_counts": {
    "in_progress": 0,
    "completed": 0,
    "failed": 0,
    "cancelled": 0,
    "total": 0
  },
  "metadata": {},
  "expires_after": null,
  "expires_at": null,
  "last_active_at": 1718321259
}
```

Reply:
So this is not something we can follow in proto/gRPC as we cannot distinguish unset v.s. explicitly setting to null.

Reply:
ccing @Ladicle as she might have some idea.

Reply:
maybe we can use `google.protobuf.Int32Value` in "google/protobuf/wrappers.proto" package instead

Reply:
`Int32Value ` is an option to differentiate the clearing vs unsetting, but it may make our API be incompatible with openAI's API. Added a TODO comments to address it later.